6.1 Alignment objects

6.1.1 Creating an Alignment object from sequences and coordinates

In [1]:
seqA = "CCGGTTTTT"
seqB = "AGTTTAA"
seqC = "AGGTTT"
sequences = [seqA, seqB, seqC]

In [2]:
import numpy as np

coordinates = np.array([[1, 3, 4, 7, 9], [0, 2, 2, 5, 5], [0, 2, 3, 6, 6]])

These coordinates define the alignment for the following sequence segments:

• SeqA[1:3], SeqB[0:2], and SeqC[0:2] are aligned to each other;

• SeqA[3:4] and SeqC[2:3] are aligned to each other, with a gap of one nucleotide in seqB;

• SeqA[4:7], SeqB[2:5], and SeqC[3:6] are aligned to each other;

• SeqA[7:9] is not aligned to seqB or seqC.

Note that the alignment does not include the first nucleotide of seqA and last two nucleotides of seqB.
Now we can create the Alignment object:

In [3]:
from Bio.Align import Alignment
alignment = Alignment(sequences, coordinates)
alignment

<Alignment object (3 rows x 8 columns) at 0x19b84110220>

In [4]:
alignment.sequences

['CCGGTTTTT', 'AGTTTAA', 'AGGTTT']

In [5]:
alignment.coordinates

array([[1, 3, 4, 7, 9],
       [0, 2, 2, 5, 5],
       [0, 2, 3, 6, 6]])

In [6]:
print(alignment)

                  1 CGGTTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



6.1.2 Creating an Alignment object from aligned sequences

In [7]:
aligned_sequences = ["CGGTTTTT", "AG-TTT--", "AGGTTT--"]
sequences = [aligened_sequence.replace("-", "") for aligened_sequence in aligned_sequences]
sequences


['CGGTTTTT', 'AGTTT', 'AGGTTT']

In [8]:
coordinates = Alignment.infer_coordinates(aligned_sequences)
coordinates

array([[0, 2, 3, 6, 8],
       [0, 2, 2, 5, 5],
       [0, 2, 3, 6, 6]])

The initial G nucleotide of seqA and the final CC nucleotides of seqB were not included in the alignment and
is therefore missing here. But this is easy to fix:

In [10]:
sequences[0] = "C" + sequences[0]
sequences[1] = sequences[1] + "AA"
sequences

['CCCGGTTTTT', 'AGTTTAAAA', 'AGGTTT']

In [11]:
coordinates[0, :] += 1
coordinates

array([[1, 3, 4, 7, 9],
       [0, 2, 2, 5, 5],
       [0, 2, 3, 6, 6]])

In [12]:
alignment = Alignment(sequences, coordinates)
print(alignment)

                  1 CCGGTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



In [13]:
ungapped_alignment = Alignment(["ACGTACGT", "AAGTACGT", "ACGTACCT"])
ungapped_alignment

<Alignment object (3 rows x 8 columns) at 0x19b84d23f40>

In [14]:
ungapped_alignment.coordinates

array([[0, 8],
       [0, 8],
       [0, 8]])

In [15]:
print(ungapped_alignment)

                  0 ACGTACGT 8
                  0 AAGTACGT 8
                  0 ACGTACCT 8



6.1.3 Common alignment attributes

6.2 Slicing and indexing an alignment

In [16]:
print(alignment)

                  1 CCGGTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



In [17]:
alignment.length

8

In [18]:
alignment[0]

'CCGGTTTT'

In [19]:
alignment[1]

'AG-TTT--'

In [20]:
alignment[2]

'AGGTTT--'

In [21]:
alignment[0, :]

'CCGGTTTT'

In [24]:
alignment[1, :]

'AG-TTT--'

In [25]:
alignment[0, 1:-1]

'CGGTTT'

In [30]:
alignment[1, 1:-1]

'G-TTT-'

In [31]:
alignment[0, (1, 2, 4)]

'CGT'

In [32]:
alignment[1, range(0, 5, 2)]

'A-T'

In [33]:
alignment[0, 2]

'G'

In [34]:
alignment[2, 6]

'-'

In [36]:
print(alignment)

                  1 CCGGTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



In [37]:
alignment[:, 0]

'CAA'

In [38]:
alignment[:, 1]

'CGG'

In [39]:
alignment[:, 2]

'G-G'

In [40]:
alignment[1:]

<Alignment object (2 rows x 6 columns) at 0x19b844292e0>

In [42]:
print(alignment[1:])

target            0 AG-TTT 5
                  0 ||-||| 6
query             0 AGGTTT 6



In [43]:
print(alignment[1:, :])

target            0 AG-TTT 5
                  0 ||-||| 6
query             0 AGGTTT 6



In [44]:
alignment[:, :4]

<Alignment object (3 rows x 4 columns) at 0x19b843d80d0>

In [45]:
print(alignment[:, :4])

                  1 CCGG 5
                  0 AG-T 3
                  0 AGGT 4



In [46]:
alignment[:, -6:]

<Alignment object (3 rows x 6 columns) at 0x19b84645b20>

In [47]:
print(alignment[:, -6:])

                  3 GGTTTT 9
                  2 -TTT-- 5
                  2 GTTT-- 6



In [49]:
print(alignment[:, (1, 3, 0)])

                  0 CGC 3
                  0 GTA 3
                  0 GTA 3



In [ ]:
alignment[:, :]

6.3 Getting information about the alignment

6.3.1 Alignment shape

In [50]:
len(alignment)

3

In [51]:
alignment.length

8

In [52]:
alignment.shape

(3, 8)

6.3.2 Comparing alignments


Two alignments are equal to each other (meaning that alignment1 == alignment2 evaluates to True)
if each of the sequences in alignment1.sequences and alignment2.sequences are equal to each other,
and alignment1.coordinates and alignment2.coordinates contain the same coordinates. If either of
these conditions is not fulfilled, then alignment1 == alignment2 evaluates to False. Inequality of two
alignments (e.g., alignment1 < alignment2) is established by first comparing alignment1.sequences and
alignment2.sequences, and if they are equal, by comparing alignment1.coordinates to alignment2.coordinates

6.3.3 Finding the indices of aligned sequences

In [54]:
pairwise_alignment = alignment[:2, :]
print(pairwise_alignment)

target            1 CCGGTTTT 9
                  0 ..-.||-- 8
query             0 AG-TTT-- 5



In [55]:
pairwise_alignment.aligned

array([[[1, 3],
        [4, 7]],

       [[0, 2],
        [2, 5]]])

In [57]:
pairwise_alignment1 = Alignment(["AAACAAA", "AAAGAAA"],
                                np.array([[0, 3, 4, 4, 7], [0, 3, 3, 4, 7]]))
pairwise_alignment2 = Alignment(["AAACAAA", "AAAGAAA"],
                                np.array([[0, 3, 3, 4, 7], [0, 3, 4, 4, 7]]))

In [58]:
print(pairwise_alignment1)

target            0 AAAC-AAA 7
                  0 |||--||| 8
query             0 AAA-GAAA 7



In [59]:
print(pairwise_alignment2)

target            0 AAA-CAAA 7
                  0 |||--||| 8
query             0 AAAG-AAA 7



In [60]:
pairwise_alignment1.aligned

array([[[0, 3],
        [4, 7]],

       [[0, 3],
        [4, 7]]])

In [61]:
pairwise_alignment2.aligned

array([[[0, 3],
        [4, 7]],

       [[0, 3],
        [4, 7]]])

In [62]:
print(alignment)

                  1 CCGGTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



In [63]:
alignment.indices

array([[ 1,  2,  3,  4,  5,  6,  7,  8],
       [ 0,  1, -1,  2,  3,  4, -1, -1],
       [ 0,  1,  2,  3,  4,  5, -1, -1]])

In [64]:
alignment.sequences

['CCCGGTTTTT', 'AGTTTAAAA', 'AGGTTT']

In [65]:
alignment.inverse_indices

[array([-1,  0,  1,  2,  3,  4,  5,  6,  7, -1]),
 array([ 0,  1,  3,  4,  5, -1, -1, -1, -1]),
 array([0, 1, 2, 3, 4, 5])]

6.3.4 Counting identities, mismatches, and gaps

In [66]:
print(pairwise_alignment)

target            1 CCGGTTTT 9
                  0 ..-.||-- 8
query             0 AG-TTT-- 5



In [67]:
pairwise_alignment.counts()

AlignmentCounts(gaps=3, identities=2, mismatches=3)

In [69]:
print(alignment)

                  1 CCGGTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



In [70]:
alignment.counts()

AlignmentCounts(gaps=8, identities=10, mismatches=6)

6.3.5 Letter frequencies

In [71]:
alignment.frequencies

{'C': array([1., 1., 0., 0., 0., 0., 0., 0.]),
 'G': array([0., 2., 2., 1., 0., 0., 0., 0.]),
 'T': array([0., 0., 0., 2., 3., 3., 1., 1.]),
 'A': array([2., 0., 0., 0., 0., 0., 0., 0.]),
 '-': array([0., 0., 1., 0., 0., 0., 2., 2.])}

6.3.6 Substitutions

In [73]:
m = alignment.substitutions
print(m)

    A   C   G   T
A 1.0 0.0 0.0 0.0
C 2.0 0.0 2.0 0.0
G 0.0 0.0 2.0 2.0
T 0.0 0.0 0.0 7.0



In [74]:
m["C", "A"]

2.0

In [76]:
m["A", "C"]

0.0

In [77]:
m += m.transpose()
m /= 2.0
print(m)

    A   C   G   T
A 1.0 1.0 0.0 0.0
C 1.0 0.0 1.0 0.0
G 0.0 1.0 2.0 1.0
T 0.0 0.0 1.0 7.0



In [78]:
m["A", "C"]

1.0

In [79]:
m["C", "A"]

1.0

6.3.7 Alignments as arrays

In [80]:
align_array = np.array(alignment)
align_array.shape

(3, 8)

In [81]:
align_array

array([[b'C', b'C', b'G', b'G', b'T', b'T', b'T', b'T'],
       [b'A', b'G', b'-', b'T', b'T', b'T', b'-', b'-'],
       [b'A', b'G', b'G', b'T', b'T', b'T', b'-', b'-']], dtype='|S1')

In [82]:
align_array = np.array(alignment, dtype="U")
align_array

array([['C', 'C', 'G', 'G', 'T', 'T', 'T', 'T'],
       ['A', 'G', '-', 'T', 'T', 'T', '-', '-'],
       ['A', 'G', 'G', 'T', 'T', 'T', '-', '-']], dtype='<U1')

6.4 Operations on an alignment

In [83]:
print(alignment)

                  1 CCGGTTTT 9
                  0 AG-TTT-- 5
                  0 AGGTTT-- 6



In [84]:
alignment.sort()
print(alignment)

                  0 AGGTTT-- 6
                  0 AG-TTT-- 5
                  1 CCGGTTTT 9



In [85]:
from Bio.SeqUtils import gc_fraction
alignment.sort(key=gc_fraction)
print(alignment)

                  0 AG-TTT-- 5
                  0 AGGTTT-- 6
                  1 CCGGTTTT 9



In [86]:
alignment.sort(key=gc_fraction, reverse=True)
print(alignment)

                  1 CCGGTTTT 9
                  0 AGGTTT-- 6
                  0 AG-TTT-- 5



6.4.2 Reverse-complementing the alignment

In [87]:
alignment.sequences

['CCCGGTTTTT', 'AGGTTT', 'AGTTTAAAA']

In [88]:
rc_alignment = alignment.reverse_complement()
print(rc_alignment.sequences)

['AAAAACCGGG', 'AAACCT', 'TTTTAAACT']


In [90]:
print(rc_alignment)

                  1 AAAACCGG 9
                  0 --AAACCT 6
                  4 --AAA-CT 9



In [91]:
alignment[:, :4].sequences

['CCCGGTTTTT', 'AGGTTT', 'AGTTTAAAA']

In [92]:
print(alignment[:, :4])

                  1 CCGG 5
                  0 AGGT 4
                  0 AG-T 3



In [94]:
rc_alignment = alignment[:, :4].reverse_complement()
rc_alignment[:, :4].sequences

['AAAAACCGGG', 'AAACCT', 'TTTTAAACT']

In [95]:
print(rc_alignment[:, :4])

                  5 CCGG 9
                  2 ACCT 6
                  6 A-CT 9



6.4.3 Adding alignments

In [104]:
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
a1 = SeqRecord(Seq("AAAAC"), id="Alpha")
b1 = SeqRecord(Seq("AAAC"), id="Beta")
c1 = SeqRecord(Seq("AAAAG"), id="Gamma")
a2 = SeqRecord(Seq("GTT"), id="Alpha")
b2 = SeqRecord(Seq("TT"), id="Beta")
c2 = SeqRecord(Seq("GT"), id="Gamma")
left = Alignment(
    [a1, b1, c1], coordinates=np.array([[0, 3, 4, 5], [0, 3, 3, 4], [0, 3, 4, 5]])
)

left.annotations = {"tool": "demo", "name": "start"}
left.column_annotations = {"stats": "CCCXC"}

right = Alignment(
    [a2, b2, c2], coordinates=np.array([[0, 1, 2, 3], [0, 0, 1, 2], [0, 1, 1, 2]])
)
right.annotations = {"tool": "demo", "name": "end"}
right.column_annotations = {"stats": "CXC"}

In [103]:
print(left)

Alpha             0 AAAAC 5
Beta              0 AAA-C 4
Gamma             0 AAAAG 5



In [106]:
print(right)

Alpha             0 GTT 3
Beta              0 -TT 2
Gamma             0 G-T 2



In [107]:
combined = left + right
print(combined)

Alpha             0 AAAACGTT 8
Beta              0 AAA-C-TT 6
Gamma             0 AAAAGG-T 7



In [108]:
len(left)

3

In [109]:
len(right)

3

In [110]:
len(combined)

3

In [111]:
combined.annotations

{'tool': 'demo'}

In [112]:
combined.column_annotations

{'stats': 'CCCXCCXC'}

6.4.4 Mapping a pairwise sequence alignment

In [113]:
chromosome = "AAAAAAAACCCCCCCAAAAAAAAAAAGGGGGGAAAAAAAA"
transcript = "CCCCCCCGGGGGG"
sequences1 = [chromosome, transcript]
coordinates1 = np.array([[8, 15, 26, 32], [0, 7, 7, 13]])
alignment1 = Alignment(sequences1, coordinates1)
print(alignment1)

target            8 CCCCCCCAAAAAAAAAAAGGGGGG 32
                  0 |||||||-----------|||||| 24
query             0 CCCCCCC-----------GGGGGG 13



In [114]:
rnaseq = "CCCCGGGG"
sequences2 = [transcript, rnaseq]
coordinates2 = np.array([[3, 11], [0, 8]])
alignment2 = Alignment(sequences2, coordinates2)
print(alignment2)

target            3 CCCCGGGG 11
                  0 ||||||||  8
query             0 CCCCGGGG  8



In [115]:
alignment3 = alignment1.map(alignment2)
print(alignment3)

target           11 CCCCAAAAAAAAAAAGGGG 30
                  0 ||||-----------|||| 19
query             0 CCCC-----------GGGG  8



In [116]:
alignment3.coordinates

array([[11, 15, 26, 30],
       [ 0,  4,  4,  8]])

In [117]:
format(alignment3, "psl")

'8\t0\t0\t0\t0\t0\t1\t11\t+\tquery\t8\t0\t8\ttarget\t40\t11\t30\t2\t4,4,\t0,4,\t11,26,\n'

In [119]:
from Bio import Align
chain = Align.read("panTro5ToPanTro6.over.chain", "chain")
chain.sequences[0].id

'chr1'

In [120]:
len(chain.sequences[0].seq)

228573443

In [121]:
chain.sequences[1].id

'chr1'

In [124]:
len(chain.sequences[1].seq)

224244399

In [125]:
import numpy as np
np.set_printoptions(threshold=5) # print 5 array elements per row
print(chain.coordinates) # doctest:+ELLIPSIS

[[122250000 122250400 122250400 ... 122909818 122909819 122909835]
 [111776384 111776784 111776785 ... 112019962 112019962 112019978]]


In [127]:
transcript = Align.read("est.panTro5.psl", "psl")
transcript.sequences[0].id
# 'chr1'

'chr1'

In [128]:
len(transcript.sequences[0].seq)

228573443

In [130]:

transcript.sequences[1].id

'DC525629'

In [131]:
len(transcript.sequences[1].seq)

407

In [133]:
print(transcript.coordinates)

[[122835789 122835847 122840993 122841145 122907212 122907314]
 [       32        90        90       242       242       344]]


In [134]:
len(chain.sequences[0].seq) == len(transcript.sequences[0].seq)

True

In [135]:
chain = chain[::-1]
chain.sequences[0].id

'chr1'

In [136]:
len(chain.sequences[0].seq)

224244399

In [137]:
chain.sequences[1].id

'chr1'

In [138]:
len(chain.sequences[1].seq)

228573443

In [139]:
print(chain.coordinates)

[[111776384 111776784 111776785 ... 112019962 112019962 112019978]
 [122250000 122250400 122250400 ... 122909818 122909819 122909835]]


In [140]:
np.set_printoptions(threshold=1000)

Now we can get the coordinates of DC525629 against chimpanzee genome assembly panTro6 by calling
chain.map, with transcript as the argumen

In [141]:
lifted_transcript = chain.map(transcript)
lifted_transcript.sequences[0].id

'chr1'

In [142]:
len(lifted_transcript.sequences[0].seq)

224244399

In [143]:
lifted_transcript.sequences[1].id

'DC525629'

In [144]:
len(lifted_transcript.sequences[1].seq)

407

In [145]:
 print(lifted_transcript.coordinates)

[[111982717 111982775 111987921 111988073 112009200 112009302]
 [       32        90        90       242       242       344]]


This shows that nucleotide range 32:344 of expressed sequence tag DC525629 aligns to range 111982717:112009302
of chr1 of chimpanzee genome assembly panTro6. Note that the genome span of DC525629 on chimpanzee genome assembly panTro5 is 122907314 - 122835789 = 71525 bp, while on panTro6 the genome span is
112009302 - 111982717 = 26585 bp.


这显示了表达序列标签DC525629的核苷酸范围32:344与黑猩猩基因组装配panTro6中chr1的范围111982717:112009302进行了比对。注意，在黑猩猩基因组装配panTro5上，DC525629的基因组跨度是122907314 - 122835789 = 71525 bp，而在panTro6上的基因组跨度是112009302 - 111982717 = 26585 bp。

6.4.5 Mapping a multiple sequence alignment

Consider a multiple alignment of genomic sequences of chimpanzee, human, macaque, marmoset, mouse,
and rat:

In [146]:
from Bio import Align
path = "panTro5.maf"
genome_alignment = Align.read(path, "maf")
for record in genome_alignment.sequences:
    print(record.id, len(record.seq))

panTro5.chr1 228573443
hg19.chr1 249250621
rheMac8.chr1 225584828
calJac3.chr18 47448759
mm10.chr3 160039680
rn6.chr2 266435125


In [147]:
genome_alignment.coordinates

array([[133922962, 133922962, 133922970, 133922970, 133922972, 133922972,
        133922995, 133922998, 133923010],
       [155784573, 155784573, 155784581, 155784581, 155784583, 155784583,
        155784606, 155784609, 155784621],
       [130383910, 130383910, 130383918, 130383918, 130383920, 130383920,
        130383943, 130383946, 130383958],
       [  9790455,   9790455,   9790463,   9790463,   9790465,   9790465,
          9790488,   9790491,   9790503],
       [ 88858039,  88858036,  88858028,  88858026,  88858024,  88858020,
         88857997,  88857997,  88857985],
       [188162970, 188162967, 188162959, 188162959, 188162957, 188162953,
        188162930, 188162930, 188162918]])

In [148]:
print(genome_alignment)

panTro5.c 133922962 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc
hg19.chr1 155784573 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc
rheMac8.c 130383910 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc
calJac3.c   9790455 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAGCTTAAAggct
mm10.chr3  88858039 TATAATAATTGTATATGTCACAGAAAAAAATGAATTTTCAAT---GACTTAATAGCC
rn6.chr2  188162970 TACAATAATTG--TATGTCATAGAAAAAAATGAATTTTCAAT---AACTTAATAGCC

panTro5.c 133923010
hg19.chr1 155784621
rheMac8.c 130383958
calJac3.c   9790503
mm10.chr3  88857985
rn6.chr2  188162918



In [149]:
paths = [
"panTro5ToPanTro6.chain",
"hg19ToHg38.chain",
"rheMac8ToRheMac10.chain",
"calJac3ToCalJac4.chain",
"mm10ToMm39.chain",
"rn6ToRn7.chain",
]
liftover_alignments = [Align.read(path, "chain") for path in paths]
for liftover_alignment in liftover_alignments:
    print(liftover_alignment.target.id, liftover_alignment.coordinates[0, :])

chr1 [133919957 133924947 133924947 133926309 133926312 133932620]
chr1 [155184381 156354347 156354348 157128497 157128497 157137496]
chr1 [130382477 130383872 130383872 130384222 130384222 130388520]
chr18 [9786631 9787941 9788508 9788508 9795062 9795065 9795737]
chr3 [66807541 74196805 74196831 94707528 94707528 94708176 94708178 94708718]
chr2 [188111581 188158351 188158351 188171225 188171225 188228261 188228261
 188236997]


Note that the order of species is the same in liftover_alignments and genome_alignment.sequences.
Now we can lift over the multiple sequence alignment to the new genome assembly versions

In [150]:
genome_alignment = genome_alignment.mapall(liftover_alignments)
for record in genome_alignment.sequences:
    print(record.id, len(record.seq))

chr1 224244399
chr1 248956422
chr1 223616942
chr18 47031477
chr3 159745316
chr2 249053267


In [151]:
genome_alignment.coordinates

array([[130611000, 130611000, 130611008, 130611008, 130611010, 130611010,
        130611033, 130611036, 130611048],
       [155814782, 155814782, 155814790, 155814790, 155814792, 155814792,
        155814815, 155814818, 155814830],
       [ 95186253,  95186253,  95186245,  95186245,  95186243,  95186243,
         95186220,  95186217,  95186205],
       [  9758318,   9758318,   9758326,   9758326,   9758328,   9758328,
          9758351,   9758354,   9758366],
       [ 88765346,  88765343,  88765335,  88765333,  88765331,  88765327,
         88765304,  88765304,  88765292],
       [174256702, 174256699, 174256691, 174256691, 174256689, 174256685,
        174256662, 174256662, 174256650]])

In [ ]:
from Bio import SeqIO
names = ("panTro6", "hg38", "rheMac10", "calJac4", "mm39", "rn7")
for i, name in enumerate(names):
    filename = f"{name}.2bit"
    genome = SeqIO.parse(filename, "twobit")
    chromosome = genome_alignment.sequences[i].id
    assert len(genome_alignment.sequences[i]) == len(genome[chromosome])
    genome_alignment.sequences[i] = genome[chromosome]
    genome_alignment.sequences[i].id = f"{name}.{chromosome}"

 print(genome_alignment)

# panTro6.c 130611000 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc
# hg38.chr1 155814782 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc
# rheMac10. 95186253 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc
# calJac4.c 9758318 ---ACTAGTTA--CA----GTAACAGAaaataaaatttaaatagaagcttaaaggct
# mm39.chr3 88765346 TATAATAATTGTATATGTCACAGAAAAAAATGAATTTTCAAT---GACTTAATAGCC
# rn7.chr2 174256702 TACAATAATTG--TATGTCATAGAAAAAAATGAATTTTCAAT---AACTTAATAGCC
# <BLANKLINE>
# panTro6.c 130611048
# hg38.chr1 155814830
# rheMac10. 95186205
# calJac4.c 9758366
# mm39.chr3 88765292
# rn7.chr2 174256650
# <BLANKLINE>
# The mapall method can also be used to create a multiple alignment of codon sequences from a multiple
# sequence alignment of the corresponding amino acid sequences (see Section 7.12.2 for details)

6.5 The Alignments class
The Alignments (plural) class inherits from AlignmentsAbstractBaseClass and from list, and can be
used as a list to store Alignment objects. The behavior of Alignments objects is different from that of list
objects in three important ways:

In [152]:
from Bio.Align import Alignments
alignment_list = [alignment1, alignment2, alignment3]
for item in alignment_list:
    print(repr(item)) # doctest: +ELLIPSIS

<Alignment object (2 rows x 24 columns) at 0x19b84d40880>
<Alignment object (2 rows x 8 columns) at 0x19b84d23610>
<Alignment object (2 rows x 19 columns) at 0x19b840ff940>


In [153]:
for item in alignment_list:
    print(repr(item)) # doctest: +ELLIPSIS

<Alignment object (2 rows x 24 columns) at 0x19b84d40880>
<Alignment object (2 rows x 8 columns) at 0x19b84d23610>
<Alignment object (2 rows x 19 columns) at 0x19b840ff940>


In [154]:
alignments = Alignments([alignment1, alignment2, alignment3])
for item in alignments:
    print(repr(item)) # doctest: +ELLIPSIS

<Alignment object (2 rows x 24 columns) at 0x19b84d40880>
<Alignment object (2 rows x 8 columns) at 0x19b84d23610>
<Alignment object (2 rows x 19 columns) at 0x19b840ff940>


In [156]:
for item in alignments:
    print(repr(item)) # doctest: +ELLIPSIS

In [157]:
alignments.rewind()
for item in alignments:
    print(repr(item)) # doctest: +ELLIPSIS

<Alignment object (2 rows x 24 columns) at 0x19b84d40880>
<Alignment object (2 rows x 8 columns) at 0x19b84d23610>
<Alignment object (2 rows x 19 columns) at 0x19b840ff940>


In [158]:
alignment_list.score = 100 # doctest: +ELLIPSIS

AttributeError: 'list' object has no attribute 'score'

In [160]:
alignments.score = 100
alignments.score

100

6.6 Reading and writing alignments

In [161]:
from Bio import Align
target = "myfile.txt"
Align.write(alignments, target, "clustal")

0

In [162]:
from Bio import Align
alignments = Align.Alignments(alignments)
metadata = {"Program": "Biopython", "Version": "1.81"}
alignments.metadata = metadata
target = "myfile.txt"
Align.write(alignments, target, "clustal")

0

In [163]:
str(alignment)

'                  1 CCGGTTTT 9\n                  0 AGGTTT-- 6\n                  0 AG-TTT-- 5\n'

In [164]:
format(alignment)

'                  1 CCGGTTTT 9\n                  0 AGGTTT-- 6\n                  0 AG-TTT-- 5\n'

In [165]:
print(format(alignment))

                  1 CCGGTTTT 9
                  0 AGGTTT-- 6
                  0 AG-TTT-- 5



In [166]:
format(alignment, "clustal")

'sequence_0                          CCGGTTTT\nsequence_1                          AGGTTT--\nsequence_2                          AG-TTT--\n\n\n'

In [167]:
print(format(alignment, "clustal"))

sequence_0                          CCGGTTTT
sequence_1                          AGGTTT--
sequence_2                          AG-TTT--





In [168]:
print(f"*** this is the alignment in Clustal format: ***\n{alignment:clustal}\n***")

*** this is the alignment in Clustal format: ***
sequence_0                          CCGGTTTT
sequence_1                          AGGTTT--
sequence_2                          AG-TTT--



***


In [169]:
format(alignment, "maf")

'a\ns sequence_0 1 8 + 10 CCGGTTTT\ns sequence_1 0 6 +  6 AGGTTT--\ns sequence_2 0 5 +  9 AG-TTT--\n\n'

In [170]:
print(format(alignment, "maf"))

a
s sequence_0 1 8 + 10 CCGGTTTT
s sequence_1 0 6 +  6 AGGTTT--
s sequence_2 0 5 +  9 AG-TTT--




In [171]:
print(pairwise_alignment)

target            1 CCGGTTTT 9
                  0 ..-.||-- 8
query             0 AG-TTT-- 5



In [172]:
print(format(pairwise_alignment, "bed"))

target	1	7	query	0	+	1	7	0	2	2,3,	0,3,



In [173]:
print(pairwise_alignment.format("bed"))

target	1	7	query	0	+	1	7	0	2	2,3,	0,3,



In [174]:
print(pairwise_alignment.format("bed", bedN=3))

target	1	7



In [175]:
print(pairwise_alignment.format("bed", bedN=6))

target	1	7	query	0	+



In [176]:
from Bio import Align
alignment = Align.read("probcons.fa", "fasta")
alignment

<Alignment object (5 rows x 101 columns) at 0x19b85dd3880>

In [177]:
print(alignment)

plas_horv         0 D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQE
plas_chlr         0 --VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRD
plas_anav         0 --VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHK
plas_proh         0 VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNT
azup_achc         0 VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A------

plas_horv        57 EYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV 95
plas_chlr        56 DYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV 94
plas_anav        58 QLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV 99
plas_proh        56 KLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV 94
azup_achc        51 -FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV 88



In [178]:
print(format(alignment, "fasta"))

>plas_horvu
D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV
>plas_chlre
--VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV
>plas_anava
--VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV
>plas_proho
VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV
>azup_achcy
VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV



In [180]:
print(format(alignment, "clustal"))

plas_horvu                          D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-
plas_chlre                          --VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-
plas_anava                          --VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKS
plas_proho                          VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-
azup_achcy                          VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-

plas_horvu                          VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVT
plas_chlre                          VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKII
plas_anava                          ADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKIT
plas_proho                          ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTIT
azup_achcy                          AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVE

plas_horvu                          V
plas_chlre                          V
plas_anava                          V
plas_proho    

In [181]:
alignment.sequences

[SeqRecord(seq=Seq('DVLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSGVDVSKI...VTV'), id='plas_horvu', name='<unknown name>', description='', dbxrefs=[]),
 SeqRecord(seq=Seq('VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSGVNADAIS...IIV'), id='plas_chlre', name='<unknown name>', description='', dbxrefs=[]),
 SeqRecord(seq=Seq('VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKS...ITV'), id='plas_anava', name='<unknown name>', description='', dbxrefs=[]),
 SeqRecord(seq=Seq('VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDKVPAGESAPALS...ITV'), id='plas_proho', name='<unknown name>', description='', dbxrefs=[]),
 SeqRecord(seq=Seq('VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDKGHNVETIKGMIPDGAEAFKS...VEV'), id='azup_achcy', name='<unknown name>', description='', dbxrefs=[])]

In [182]:
alignment.coordinates

array([[ 0,  1,  1, 33, 34, 42, 44, 48, 48, 50, 50, 51, 58, 73, 73, 95],
       [ 0,  0,  0, 32, 33, 41, 43, 47, 47, 49, 49, 50, 57, 72, 72, 94],
       [ 0,  0,  0, 32, 33, 41, 43, 47, 48, 50, 51, 52, 59, 74, 77, 99],
       [ 0,  1,  2, 34, 35, 43, 43, 47, 47, 49, 49, 50, 57, 72, 72, 94],
       [ 0,  1,  2, 34, 34, 42, 44, 48, 48, 50, 50, 51, 51, 66, 66, 88]])

In [183]:
from io import StringIO
stream = StringIO()
Align.write(alignment, stream, "FASTA")


1

In [184]:
print(stream.getvalue())

>plas_horvu
D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV
>plas_chlre
--VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV
>plas_anava
--VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV
>plas_proho
VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV
>azup_achcy
VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV



In [186]:
from Bio import Align
alignments = Align.parse("opuntia.aln", "clustal")

In [187]:
alignments.metadata

{'Program': 'CLUSTAL', 'Version': '2.1'}

In [188]:
alignment = next(alignments)

In [189]:
print(alignment)

gi|627328         0 TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT
gi|627328         0 TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT
gi|627328         0 TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT
gi|627328         0 TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT
gi|627329         0 TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT
gi|627328         0 TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT
gi|627329         0 TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT

gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA
gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA
gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA
gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA
gi|627329        60 CTAAATGATATAGGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA
gi|627328        60 CTAAATG

: 